library-best-practices/JLBP-4.md

```diff
@@ -0,0 +1,57 @@
+[JLBP-4] Avoid dependencies on unstable libraries and features
+--------------------------------------------------------------
+
+- At a high level, don't depend on things that can have frequent breaking
+  changes (where frequent means more than once every couple years).
```

Comment:
Do you think there is a concrete change I need to make here?

Reply:
I guess the main thing is that this seems too binary. "Oh, it makes breaking changes every 3 years, so it's fine to depend on" sounds like a recipe for pain, especially when multiplied by _n_ dependencies.

The main message I'd send is probably:

- Don't depend on anything that makes breaking changes unless you have a very good reason.
- If you think you need to depend on something that makes breaking changes, try advocating for them to stop. I have some experience being the offender here, so I can offer some concrete tips:
  - If necessary and practical, provide resources to help them reach the point at which they're comfortable stopping.
  - Maybe show them data on the changes they've made historically to help them judge how important breaking changes are to them nowadays.
  - Maybe ask us for horror stories about how people are never going to upgrade because it's hard enough to coordinate major-version bumps within a closed ecosystem, let alone across ecosystems.
- If you do need to depend on something that makes breaking changes, then at least try to get its owners to deprecate things well in advance of removal (and maybe batch together the removals into fewer, larger breaking changes, too).

Reply:
Hmm you seem to be talking about advocacy for robust practices, which seems to be outside the original scope of this JLBP, which is just to avoid dependencies on libraries that don't follow robust practices. There is a candidate best practice (JLBP-10 - Infrequent major version bumps) that I think covers your idea here (and should probably incorporate your specific points). Could we get by with a much more limited adjustment for the purposes of the present JLBP? Maybe:

"At a high level, don't depend on things that can have frequent or unnecessary breaking changes. Here, 'frequent' means more than once every couple years, and 'unnecessary' means breaking features that aren't causing problems."

Reply:
Sorry that I continue to be so scattered about this -- and that I'm thinking through much of this only now :(

> Hmm you seem to be talking about advocacy for robust practices, which seems to be outside the original scope of this JLBP, which is just to avoid dependencies on libraries that don't follow robust practices.

Yes and no. I do agree that the details don't belong in this page. What I think does belong here is some encouragement to users to advocate when possible. It was apparently very easy for people to go "Well, Guava's unstable, but I like Guava, so I'll just use it," as if "talk to the Guava team" was some foreign, non-engineering task. OK, that's not fair: We did of course have a number of people talk to us, though often not in the most helpful ways -- hence all the details I mentioned above, which yeah, should wait for JLBP-10. I'm not trying to make this out to be easy, and ultimately the main problem was us on the Guava team, but someone could probably have helped us along earlier, and that may well have had greater impact than whatever else that person was doing at the time. (To reiterate, the main problem was still our fault! I just additionally place a small amount of fault on people who behaved as if they had no choice but to depend on an unstable API.)

But anyway, that's only one option. The other two things I would suggest for here are:

1. It's possible to shade or fork. We probably don't want to get into that deeply, but it's an option if you really do need to use something unstable. (Maybe gRPC did this with Netty or something??? I may be thinking of the wrong example.)

2. I think "more than once every couple years" is giving people permission to depend on libraries that break every few years, and I think that's going to cause problems. I would probably concede that less frequent breakages are better -- but I could Devil's Advocate for the opposite, arguing that it's better to depend on something that breaks frequently so that people see the risks in action and can't minimize them :) -- but I don't think we should give a number that suggests "no problem if you clear this bar; problem if you don't."

(It's possible that I'm too focused on the Guava case here. Are there other experiences with major-version updates that you've had that I should be thinking about in addition to Guava?)

Reply:
(Maybe the summary here is: "The Guava team made bad decisions. Here's how you can cope with them so that you don't end up in the blast radius.")

Reply:
I understand why you want the advocacy - I want it too. Maybe we could even create a JLBP for it - "advocate for best practices with your dependencies." However, it's a meta-action. Anyway, these JLBP's may become a launch requirement down the road, such that for any new libraries, no advocacy is required. 

1. Yes, grpc did create a variant of grpc-netty called grpc-netty-shaded. (As an aside, having both doesn't create conflicts, because they have a specific design that still works when both are present.) For shading in general, I'm uneasy about recommending it because of all the drawbacks it has (such as the fact that size can really explode). I agree that we shouldn't go deeply into that here; I think I want to create a JLBP later on the topic of shading (which will generally recommend against it). I'd honestly want someone to read the present JLBP, think "hey, now I can't do X," then talk to us, as opposed to the scenario where we recommend something with difficult tradeoffs, and have it over-applied. You could argue that having a user shade a dependency is still complying with the rules here anyway (even if we don't mention it), because it's no longer an artifact dependency (since it's bundled). 

2. If we don't give a time frame, then "frequent" will be left up to people's imagination. Some people might interpret that as "once every month," and conclude that it's fine to depend on something that only breaks surface once per year. I think saying once every five years is much too conservative for some cases (e.g. API clients, where the underlying APIs themselves can actually deprecate things in a year's time frame). I was trying to qualify the "frequent" thing by saying that it also needed to be the case that a library wouldn't break surface "unnecessarily," such that a library that technically didn't break "frequently" (say they only break once very two years) would still not be suitable if those breakages were just surface refinements and not really necessary. Let me know if that qualification doesn't come across strong enough, and let me know if that requirement still isn't sufficiently satisfying.

Reply:
[comment moved] Seeing the comment by @briandealwis , and thinking about frequency: Maybe the right thing to say is "Don't depend on something that can make breaking changes more frequently than the cadence you want for your own library." That covers the core reason for the rule, and leaves open the exact frequency to be elaborated in another best practice (candidate: JLBP-10).

Reply:
> I'd honestly want someone to read the present JLBP, think "hey, now I can't do X," then talk to us, as opposed to the scenario where we recommend something with difficult tradeoffs, and have it over-applied.

I think that's a good principle. I think, though, that "it's OK if the API breaks only every couple years" is itself an example of a recommendation with difficult tradeoffs. More on that in a second.

> I think saying once every five years is much too conservative for some cases (e.g. API clients, where the underlying APIs themselves can actually deprecate things in a year's time frame)

Thanks for bringing this back up. I know you've mentioned it before, but I default to thinking in terms of Guava, not the API libraries.

> "Don't depend on something that can make breaking changes more frequently than the cadence you want for your own library."

I like this way of approaching it, although I think this exact phrasing is still a little optimistic.

If I want to be stable for 2 years, and I depend on a library that's stable for 2 years, I'm probably not giving 2 years of stability to my users because the two schedules likely aren't synchronized. I'm giving more like 1 year on average. (As usual, users can probably use my library for longer than that, since the specific breakages might not affect them.) And that's if I have one such dependency: If I've got 3 dependencies, we're talking more like 6 months. And of course users might take a few months to adopt my new versions, so maybe they get only ~3 months, rather than the ~21 months I'd expect if I subtract 3 months from my "2-year" guarantee.

So the full story here involves some math that isn't really _that_ hard (something like "add up the number of breakage events per year") but is more complex than `min(myStability, dep1Stability, dep2Stability, ...)`. I do very much like the idea that, say, gRPC and google-cloud-vision may have different tolerances for breakages, so I think we should include that. But if we want to give a rough mathematical rule, I'd say that you're _adding_ all your dependencies' breakage rates to your own, not taking the min. If you break things frequently, anyway, then that's dominated by your own rate, and you don't need to worry much about your dependencies. But if you're gRPC, even if you're OK with breaking every 5 years, you don't want Guava, GSON, and OpenCensus to all be breaking every 5 years, too, unless you can synchronize those schedules.

Discussion-starter (discussion-continuer? :)) proposal: "Roughly speaking, your library can breaks its users every time it makes a breaking change plus every time one of its dependencies makes a breaking change. If you think you need to depend on a library but its breakage rate makes you nervous, contact Foo-Team to discuss options."

Reply:
Yeah, that's a good point about the non-alignment of release cycles. I'll work on some wording and push an edit.